docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+``result_callback``, ``cache`` and ``executor`` (the last two are ``None`` by
+default) and initiates a coala session.
+
+* The ``bears`` argument contains the list of bears to be run for a coala
+  session.
+* The ``result_callback`` is a function that is called when the results from a
+  bear run are available.
+* The ``cache`` argument if provided enables caching and runs the session using
+  the cache provided to store the bear results. The default value of this
+  parameter is ``None`` which when provided, runs coala without a cache.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+* The new bears always have to constructed with ``section`` and ``file_dict``
+  as parameters, this is important as it makes the new dependency system
+  possible.
+* Bear writers need to keep this in mind while writing their ``__init__``
+  methods, default parameters are allowed as many as wanted but the same isn't
+  true for positional parameters (default parameters are discouraged, since the
+  dependency system automatically instantiates bears, which might lead to
+  unwanted side-effects).
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files. Bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs.
+
+Task Objects
+------------
+
+These are the representation of tasks performed by bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the ``analyze`` (even to the ``execute_task``) function.
+The task objects are instrumental to the caching mechanism as their hash values
+are stored in the cache along with the bear results and are looked up during
+each coala run to fetch the results.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+        [(filename, file), {'preferred_quotation': ""}],
+    ]
+
+Improved dependency system
+--------------------------
+
+The NextGen-Core introduces a better dependency management system than the
+one used by the old core. It features the following improvements:
+
+* A class ``DependencyTracker`` manages this task of dependency management.
+  Dependencies are added and resolved by this class and it checks for
+  circular dependencies.
+* Dependency relations between two objects are tracked using a directed
+  graph. When two objects (or nodes) are connected with a directed edge they
+  form a dependency relation.
```

Comment:
I think that should stay in the docstrings rather than the documentation the bear writers need not worry about the different directions between different bear and dependency nodes, the new dependency management takes care of that :wink:

Reply:
Well in that case how about removing the line `When two objects (or nodes) are connected with a directed edge they form a dependency relation.` altogether? 
Just mentioning *connected* in case of a DAG looks a bit incomplete to me. Its like specifying velocity of an object without mentioning the direction :stuck_out_tongue_winking_eye:

Reply:
As I had mentioned again your reviews our pointing out things that were there before but were changed due to **several iterations of much more thorough reviews** by Makman. IIRC Makman had asked me to explain in **brief** how dependency relation works.

Reply:
Hm though I agree that going into the internal details (like the graph itself) is an implementation detail. It's a bit on the edge here, mentioning the idea of a graph here makes sense, that makes things more understandable, but it might be already too much. I'm rereading the paragraph again :+1: